utils/src/main/java/com/cloud/utils/nio/Link.java

```diff
@@ -453,115 +449,150 @@ public static SSLContext initSSLContext(boolean isClient) throws GeneralSecurity
-                    if (s_logger.isTraceEnabled()) {
-                        s_logger.trace("SSL: Buffer underflowed, getting more packets");
+                    break;
+                case NEED_WRAP:
+                    myNetData.clear();
+                    try {
+                        result = sslEngine.wrap(myAppData, myNetData);
+                        handshakeStatus = result.getHandshakeStatus();
+                    } catch (SSLException sslException) {
+                        s_logger.error("SSL error occurred while processing wrap data: " + sslException.getMessage());
+                        sslEngine.closeOutbound();
+                        handshakeStatus = sslEngine.getHandshakeStatus();
+                        break;
                     }
-                    tmp_pkgBuf.clear();
-                    count = ch.read(tmp_pkgBuf);
-                    if (count == -1) {
-                        throw new IOException("Connection closed with -1 on reading size.");
+                    switch (result.getStatus()) {
+                        case OK :
+                            myNetData.flip();
+                            while (myNetData.hasRemaining()) {
+                                socketChannel.write(myNetData);
+                            }
+                            break;
+                        case BUFFER_OVERFLOW:
+                            // Will occur if there is not enough space in myNetData buffer to write all the data
+                            // that would be generated by the method wrap. Since myNetData is set to session's packet
+                            // size we should not get to this point because SSLEngine is supposed to produce messages
+                            // smaller or equal to that, but a general handling would be the following:
+                            myNetData = enlargeBuffer(myNetData, netBufferSize);
+                            break;
+                        case BUFFER_UNDERFLOW:
+                            throw new SSLException("Buffer underflow occurred after a wrap. We should not reach here.");
+                        case CLOSED:
+                            try {
+                                myNetData.flip();
+                                while (myNetData.hasRemaining()) {
+                                    socketChannel.write(myNetData);
+                                }
+                                // At this point the handshake status will probably be NEED_UNWRAP
+                                // so we make sure that peerNetData is clear to read.
+                                peerNetData.clear();
+                            } catch (Exception e) {
+                                s_logger.error("Failed to send server's CLOSE message due to socket channel's failure.");
+                                handshakeStatus = sslEngine.getHandshakeStatus();
+                            }
+                            break;
```

Comment:
This case statement is quite long.  Consider extracting to private method to increase readability/maintainability.